nifi-extension-bundles/nifi-kafka-bundle/nifi-kafka-processors/src/main/java/org/apache/nifi/kafka/processors/producer/convert/RecordStreamKafkaRecordConverter.java

```diff
@@ -0,0 +1,124 @@
+    private final ComponentLog logger;
+
+    public RecordStreamKafkaRecordConverter(
+            final RecordReaderFactory readerFactory,
+            final RecordSetWriterFactory writerFactory,
+            final HeadersFactory headersFactory,
+            final KeyFactory keyFactory,
+            final int maxMessageSize,
+            final ComponentLog logger) {
+        this.readerFactory = readerFactory;
+        this.writerFactory = writerFactory;
+        this.headersFactory = headersFactory;
+        this.keyFactory = keyFactory;
+        this.maxMessageSize = maxMessageSize;
+        this.logger = logger;
+    }
+
+    @Override
+    public Iterator<KafkaRecord> convert(
+            final Map<String, String> attributes, final InputStream in, final long inputLength)
+            throws IOException {
+        try {
+            final RecordReader reader = readerFactory.createRecordReader(attributes, in, inputLength, logger);
+            final RecordSet recordSet = reader.createRecordSet();
+            final RecordSchema schema = writerFactory.getSchema(attributes, recordSet.getSchema());
+
+            final ByteArrayOutputStream os = new ByteArrayOutputStream();
+            final RecordSetWriter writer = writerFactory.createWriter(logger, schema, os, attributes);
+            final PushBackRecordSet pushBackRecordSet = new PushBackRecordSet(recordSet);
+            return toKafkaRecordIterator(attributes, os, writer, pushBackRecordSet);
+        } catch (MalformedRecordException | SchemaNotFoundException e) {
+            throw new IOException("Stream to Record conversion failed", e);
+        }
+    }
+
+    private Iterator<KafkaRecord> toKafkaRecordIterator(
+            final Map<String, String> attributes,
+            final ByteArrayOutputStream os,
+            final RecordSetWriter writer,
+            final PushBackRecordSet pushBackRecordSet) throws IOException {
+        final ValueFactory valueFactory = new RecordValueFactory(os, writer);
+        final List<RecordHeader> headers = headersFactory.getHeaders(attributes);
+
+        return new Iterator<>() {
+            @Override
+            public boolean hasNext() {
+                try {
+                    return pushBackRecordSet.isAnotherRecord();
+                } catch (final IOException e) {
+                    throw new UncheckedIOException(e);
+                }
+            }
+
+            @Override
+            public KafkaRecord next() {
+                try {
+                    final Record record = pushBackRecordSet.next();
+                    final byte[] key = keyFactory.getKey(attributes, record);
+                    final byte[] value = valueFactory.getValue(record);
+                    ProducerUtils.checkMessageSize(maxMessageSize, value.length);
+                    return new KafkaRecord(null, null, null, key, value, headers);
+                } catch (final IOException e) {
+                    throw new UncheckedIOException(e);
```

Comment:
A message should be included:
```suggestion
                    throw new UncheckedIOException("Record conversion failed", e);
```